File: app/src/main/java/com/simpleshift/scheduler/domain/colleague_mode.kt
package com.simpleshift.scheduler.domain

import com.simpleshift.scheduler.domain.model.CommonRestResult
import com.simpleshift.scheduler.domain.model.ShiftCycleConfig
import com.simpleshift.scheduler.domain.model.ShiftType
import com.simpleshift.scheduler.domain.model.Team
import java.time.LocalDate
import java.time.temporal.ChronoUnit

fun findCommonRestDays(
    teamAId: Int,
    teamBId: Int,
    today: LocalDate = LocalDate.now(),
    daysToAnalyze: Int = 365,
    customCycle: List<ShiftType>? = null,
    referenceDate: LocalDate = ShiftCycleConfig.REFERENCE_DATE
): CommonRestResult {
    val offsetA = teamPhaseOffsetFor(teamAId, customCycle)
    val offsetB = teamPhaseOffsetFor(teamBId, customCycle)
    val teamAName = Team.ALL_TEAMS.find { it.id == teamAId }?.name ?: "班组$teamAId"
    val teamBName = Team.ALL_TEAMS.find { it.id == teamBId }?.name ?: "班组$teamBId"

    val commonDates = (0 until daysToAnalyze).mapNotNull { offset ->
        val date = today.plusDays(offset.toLong())
        val shiftA = getShiftTypeForDate(date, offsetA, customCycle, referenceDate)
        val shiftB = getShiftTypeForDate(date, offsetB, customCycle, referenceDate)
        val isRestA = shiftA == ShiftType.REST || shiftA == ShiftType.STUDY
        val isRestB = shiftB == ShiftType.REST || shiftB == ShiftType.STUDY
        if (isRestA && isRestB) date else null
    }

    val next = commonDates.firstOrNull()
    val daysUntil = next?.let { ChronoUnit.DAYS.between(today, it).toInt() }
    val todayEpoch = today.toEpochDay()
    val count30 = commonDates.count { it.toEpochDay() - todayEpoch < 30 }
    val count60 = commonDates.count { it.toEpochDay() - todayEpoch < 60 }

    return CommonRestResult(
        teamAName = teamAName,
        teamBName = teamBName,
        nextCommonRestDate = next,
        daysUntilNext = daysUntil,
        commonRestDates = commonDates,
        totalCount = commonDates.size,
        countIn30Days = count30,
        countIn60Days = count60
    )
}

File: flutter/lib/domain/algorithms/colleague_mode.dart
import '../models/common_rest_result.dart';
import '../models/shift_type.dart';
import '../models/shift_cycle_config.dart';
import '../bridge/ffi_bridge.dart';
import 'shift_calculator.dart';

/// 查找两人共同休息日 — 优先走 Rust FFI
CommonRestResult findCommonRestDays({
  required int teamAId,
  required int teamBId,
  DateTime? today,
  int daysToAnalyze = 365,
  List<ShiftType>? customCycle,
  DateTime? referenceDate,
  String Function(int)? teamNameResolver,
}) {
  final t = today ?? DateTime.now();
  final resolver = teamNameResolver ?? (int id) => 'Shift ${String.fromCharCode(64 + id)}';
  final teamAName = resolver(teamAId);
  final teamBName = resolver(teamBId);

  // Try Rust FFI
  if (customCycle == null || customCycle.length == ShiftCycleConfig.cycleLength) {
    final result = ffiGetCommonRestDays(
      teamA: teamAId, teamB: teamBId, today: t,
      daysToAnalyze: daysToAnalyze,
      referenceDate: referenceDate,
    );
    if (result != null) {
      final dates = (result['common_rest_dates'] as List?)
          ?.map((d) => DateTime.parse(d as String))
          .toList() ?? [];
      final next = result['next_common_rest'] as String?;
      return CommonRestResult(
        teamAName: teamAName,
        teamBName: teamBName,
        nextCommonRestDate: next != null ? DateTime.parse(next) : null,
        daysUntilNext: result['days_until_next'] as int?,
        commonRestDates: dates,
        totalCount: dates.length,
        countIn30Days: result['count_30_days'] as int? ?? 0,
        countIn60Days: result['count_60_days'] as int? ?? 0,
      );
    }
  }
  final offsetA = teamPhaseOffsetFor(teamAId, customCycle: customCycle);
  final offsetB = teamPhaseOffsetFor(teamBId, customCycle: customCycle);

  final commonDates = <DateTime>[];
  for (int i = 0; i < daysToAnalyze; i++) {
    final date = t.add(Duration(days: i));
    final shiftA = getShiftTypeForDate(date,
        teamPhaseOffset: offsetA,
        customCycle: customCycle,
        referenceDate: referenceDate);
    final shiftB = getShiftTypeForDate(date,
        teamPhaseOffset: offsetB,
        customCycle: customCycle,
        referenceDate: referenceDate);
    final isRestA = shiftA == ShiftType.REST || shiftA == ShiftType.STUDY;
    final isRestB = shiftB == ShiftType.REST || shiftB == ShiftType.STUDY;
    if (isRestA && isRestB) {
      commonDates.add(date);
    }
  }

  final next = commonDates.isNotEmpty ? commonDates.first : null;
  final daysUntil = next?.difference(t).inDays;
  final count30 =
      commonDates.where((d) => d.difference(t).inDays < 30).length;
  final count60 =
      commonDates.where((d) => d.difference(t).inDays < 60).length;

  return CommonRestResult(
    teamAName: teamAName,
    teamBName: teamBName,
    nextCommonRestDate: next,
    daysUntilNext: daysUntil,
    commonRestDates: commonDates,
    totalCount: commonDates.length,
    countIn30Days: count30,
    countIn60Days: count60,
  );
}
